staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go

```diff
@@ -132,49 +132,33 @@ func NonSlidingUntilWithContext(ctx context.Context, f func(context.Context), pe
 // Close stopCh to stop. f may not be invoked if stop channel is already
 // closed. Pass NeverStop to if you don't want it stop.
 func JitterUntil(f func(), period time.Duration, jitterFactor float64, sliding bool, stopCh <-chan struct{}) {
-	BackoffUntil(f, NewJitteredBackoffManager(period, jitterFactor, &clock.RealClock{}), sliding, stopCh)
+	BackoffUntil(f, Backoff{Duration: period, Jitter: jitterFactor}.Timer(), sliding, stopCh)
 }
 
-// BackoffUntil loops until stop channel is closed, run f every duration given by BackoffManager.
-//
-// If sliding is true, the period is computed after f runs. If it is false then
-// period includes the runtime for f.
-func BackoffUntil(f func(), backoff BackoffManager, sliding bool, stopCh <-chan struct{}) {
-	var t clock.Timer
-	for {
-		select {
-		case <-stopCh:
-			return
-		default:
-		}
-
-		if !sliding {
-			t = backoff.Backoff()
-		}
-
-		func() {
-			defer runtime.HandleCrash()
-			f()
-		}()
+var (
+	// RealTimer can be passed to methods that need a clock.Timer.
+	RealTimer = clock.RealClock{}.NewTimer
+)
 
-		if sliding {
-			t = backoff.Backoff()
-		}
+var (
+	// internalClock is used for test injection of clocks
+	internalClock = clock.RealClock{}
+	// internalNewTicker is used for injecting behavior into tests
+	// Deprecated: Will be removed when poller() is removed.
+	internalNewTicker func(time.Duration) *time.Ticker = time.NewTicker
+	// internalNewTimer is used for injecting behavior into tests
+	// Deprecated: Will be removed when poller() is removed.
+	internalNewTimer func(time.Duration) *time.Timer = time.NewTimer
+)
 
-		// NOTE: b/c there is no priority selection in golang
-		// it is possible for this to race, meaning we could
-		// trigger t.C and stopCh, and t.C select falls through.
-		// In order to mitigate we re-check stopCh at the beginning
-		// of every loop to prevent extra executions of f().
-		select {
-		case <-stopCh:
-			if !t.Stop() {
-				<-t.C()
-			}
-			return
-		case <-t.C():
-		}
-	}
+// BackoffUntil loops until stop channel is closed, run f every duration given by delayFn.
+// An appropriately delayFn is provided by the Backoff.Step method. If sliding is true, the
+// period is computed after f runs. If it is false then period includes the runtime for f.
+func BackoffUntil(f func(), timer Timer, sliding bool, stopCh <-chan struct{}) {
```

Comment:
May I suggest to move `f` to be the last argument? Code is more readable if `f` is a closure. And, maybe, `stopCh` should be the first argument, like a context? I assume it's not a context to be usable from the existing code?

Reply:
This PR tries to minimize the changes to existing signatures except where unavoidable.  Instead of `wait.BackoffUtil` you can use `wait.Backoff{...}.DelayFunc().Until(ctx, bool, bool, func ...)`.  BackoffUntil should probably be removed in a future release, but I didn't mark it deprecated here.